<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<title>Barcoding and gRNA Probes • karyotapR</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="96x96" href="../favicon-96x96.png">
<link rel="icon" type="”image/svg+xml”" href="../favicon.svg">
<link rel="apple-touch-icon" sizes="180x180" href="../apple-touch-icon.png">
<link rel="icon" sizes="any" href="../favicon.ico">
<link rel="manifest" href="../site.webmanifest">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.3.1/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.3.1/bootstrap.bundle.min.js"></script><link href="../deps/font-awesome-6.5.2/css/all.min.css" rel="stylesheet">
<link href="../deps/font-awesome-6.5.2/css/v4-shims.min.css" rel="stylesheet">
<script src="../deps/headroom-0.11.0/headroom.min.js"></script><script src="../deps/headroom-0.11.0/jQuery.headroom.min.js"></script><script src="../deps/bootstrap-toc-1.0.1/bootstrap-toc.min.js"></script><script src="../deps/clipboard.js-2.0.11/clipboard.min.js"></script><script src="../deps/search-1.0.0/autocomplete.jquery.min.js"></script><script src="../deps/search-1.0.0/fuse.min.js"></script><script src="../deps/search-1.0.0/mark.min.js"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Barcoding and gRNA Probes">
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>


    <nav class="navbar navbar-expand-lg fixed-top bg-light" data-bs-theme="light" aria-label="Site navigation"><div class="container">

    <a class="navbar-brand me-2" href="../index.html">karyotapR</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.0.2</small>


    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item"><a class="nav-link" href="../articles/karyotapR.html">Get started</a></li>
<li class="nav-item"><a class="nav-link" href="../reference/index.html">Reference</a></li>
<li class="active nav-item dropdown">
  <button class="nav-link dropdown-toggle" type="button" id="dropdown-articles" data-bs-toggle="dropdown" aria-expanded="false" aria-haspopup="true">Articles</button>
  <ul class="dropdown-menu" aria-labelledby="dropdown-articles">
<li><a class="dropdown-item" href="../articles/barcoding.html">Barcoding and gRNA Probes</a></li>
    <li><a class="dropdown-item" href="../articles/GMMs-for-Copy-Number-Calling.html">GMMs for Copy Number Calling</a></li>
    <li><a class="dropdown-item" href="../articles/TapExp-navigation.html">Navigating the TapestriExperiment Class</a></li>
  </ul>
</li>
<li class="nav-item"><a class="nav-link" href="../news/index.html">Changelog</a></li>
      </ul>
<ul class="navbar-nav">
<li class="nav-item"><form class="form-inline" role="search">
 <input class="form-control" type="search" name="search-input" id="search-input" autocomplete="off" aria-label="Search site" placeholder="Search for" data-search-index="../search.json">
</form></li>
<li class="nav-item"><a class="external-link nav-link" href="https://github.com/joeymays/karyotapR/" aria-label="GitHub"><span class="fa fab fa-github fa-lg"></span></a></li>
      </ul>
</div>


  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Barcoding and gRNA Probes</h1>
            
            <h4 data-toc-skip class="date">Compiled on September 16,
2023</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/joeymays/karyotapR/blob/master/vignettes/articles/barcoding.Rmd" class="external-link"><code>vignettes/articles/barcoding.Rmd</code></a></small>
      <div class="d-none name"><code>barcoding.Rmd</code></div>
    </div>

    
    
<p>This tutorial will demonstrate how to use <code>karyotapR</code> to
parse and count barcoded and gRNA reads using the specialty probes in
the KaryoTap Tapestri panels. These barcodes and gRNAs are integrated
into the cell’s genome using lentivirus and are amplified by specially
designed probes specific to the vectors. Details on the probe design can
be found in the KaryoTap publication, and details on the usage of the
functions here can be found in the package documentation and reference
on this site.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/joeymays/karyotapR" class="external-link">karyotapR</a></span><span class="op">)</span></span></code></pre></div>
<p>The procedures for counting barcoded reads and gRNA reads are largely
identical, except for pointing the function to the correct probe. Since
gRNAs sequences essentially act as barcodes, we’re going to refer to
both barcodes and gRNAs as barcodes here.</p>
<p>Barcode sequences are extracted from the aligned reads stored in the
<code>.bam</code> file that is output as part of the Tapestri Pipeline.
We’ll use a dataset (experiment 4) from the KaryoTap paper as an
example. The <code>.bam</code> file used in this tutorial can be
downloaded from the <a href="https://www.ncbi.nlm.nih.gov/sra/SRX20233247" class="external-link">SRA repository</a>.
The file needs a corresponding <code>.bai</code> index file as well. If
you did not retrieve it from the Pipeline output, you can generate a new
one (see the <code><a href="../reference/countBarcodedReads.html">countBarcodedReads()</a></code> documentation).</p>
<p>To identify the barcoded sequences within the reads, we need a lookup
table containing the barcode ID (whatever you want the name to appear as
in the result) and the corresponding nucleotide sequences for the
barcode.</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">grna.lookup</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span>ids <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gRNA1"</span>, <span class="st">"gRNA2"</span><span class="op">)</span>, sequences <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"ACGGAGGCTAAGCGTCGCAA"</span>,</span>
<span>    <span class="st">"ACTCTTGCTGTGGCATTTTC"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<p>Barcoded reads are counted using <code><a href="../reference/countBarcodedReads.html">countBarcodedReads()</a></code>
and then saved to a <code>colData</code> column named using the IDs. The
<code>.bam</code> file to use is specified by its filepath. Here we
specify the <code>probe</code> that were interested in is the “gRNA”
probe, which points the function to the probe ID stored in the
<code>grnaProbe</code> slot of the <code>TapestriExperiment</code>
object. Specifying “barcode” similarly points to the probe ID
<code>barcodeProbe</code> slot.</p>
<p>To account for PCR errors, the function by default allows for 2 base
mismatches. This can be altered to the user’s preference. Indels can be
allowed as well using <code>with.indels = TRUE</code> but is set to
<code>FALSE</code> by default.</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/slotGettersSetters.html">grnaProbe</a></span><span class="op">(</span><span class="va">tapexp</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "CO610_AMP350"</span></span>
<span><span class="fu"><a href="../reference/slotGettersSetters.html">barcodeProbe</a></span><span class="op">(</span><span class="va">tapexp</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "CO610_AMP351"</span></span></code></pre></div>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tapexp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/countBarcodedReads.html">countBarcodedReads</a></span><span class="op">(</span><span class="va">tapexp</span>, bam.file <span class="op">=</span> <span class="st">"./tapexp.bam"</span>, probe <span class="op">=</span> <span class="st">"grna"</span>, barcode.lookup <span class="op">=</span> <span class="va">grna.lookup</span>,</span>
<span>    max.mismatch <span class="op">=</span> <span class="fl">2</span>, with.indels <span class="op">=</span> <span class="cn">F</span><span class="op">)</span></span></code></pre></div>
<p>The counts for the two gRNA sequences in each cell are now associated
with the corresponding cell barcode.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">colData</span><span class="op">(</span><span class="va">tapexp</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 1825 rows and 5 columns</span></span>
<span><span class="co">#&gt;                              cell.barcode total.reads  cluster     gRNA1     gRNA2</span></span>
<span><span class="co">#&gt;                               &lt;character&gt;   &lt;numeric&gt; &lt;factor&gt; &lt;numeric&gt; &lt;numeric&gt;</span></span>
<span><span class="co">#&gt; AACAACCTACAATGTGCT-1 AACAACCTACAATGTGCT-1       27595     RPE1        18         0</span></span>
<span><span class="co">#&gt; AACAACTGGCGACCATCA-1 AACAACTGGCGACCATCA-1       33325     HCEC         0        66</span></span>
<span><span class="co">#&gt; AACAATGCAAGGATGCGT-1 AACAATGCAAGGATGCGT-1       36978     HCEC         0         0</span></span>
<span><span class="co">#&gt; AACAATGCATCGACGTTG-1 AACAATGCATCGACGTTG-1       25756     HCEC         0        34</span></span>
<span><span class="co">#&gt; AACAATGCATGAGAATCC-1 AACAATGCATGAGAATCC-1       26980     HCEC         0         0</span></span>
<span><span class="co">#&gt; ...                                   ...         ...      ...       ...       ...</span></span>
<span><span class="co">#&gt; TTGGACTTCTATCATGCT-1 TTGGACTTCTATCATGCT-1       37431     HCEC         0         0</span></span>
<span><span class="co">#&gt; TTGGAGAACACGCAAGAT-1 TTGGAGAACACGCAAGAT-1       17456     RPE1         2         0</span></span>
<span><span class="co">#&gt; TTGGTAACTACCACTAGG-1 TTGGTAACTACCACTAGG-1       69296     RPE1       150         0</span></span>
<span><span class="co">#&gt; TTGGTAACTCAATCTCCG-1 TTGGTAACTCAATCTCCG-1       17866     RPE1         5         0</span></span>
<span><span class="co">#&gt; TTGTCAACCAGACTTCGT-1 TTGTCAACCAGACTTCGT-1       31021     HPNE         0         0</span></span></code></pre></div>
<p>To quickly assign sample labels to the cells, you can use
<code>callSampleLabels()</code> to assign a label based on the gRNA
counts. This creates a colData entry specified by
<code>output.feature</code> and labels the cell with whichever of the
<code>input.feature</code>s has the highest number of counts. If there
are no counts for any input feature, the return value is
<code>NA</code>. See the documentation for ways to customize the
parameters of this process.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tapexp</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/callSampleLables.html">callSampleLables</a></span><span class="op">(</span><span class="va">tapexp</span>, input.features <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"gRNA1"</span>, <span class="st">"gRNA2"</span><span class="op">)</span>, output.feature <span class="op">=</span> <span class="st">"sample.id"</span><span class="op">)</span></span>
<span></span>
<span><span class="fu">colData</span><span class="op">(</span><span class="va">tapexp</span><span class="op">)</span></span>
<span><span class="co">#&gt; DataFrame with 1825 rows and 6 columns</span></span>
<span><span class="co">#&gt;                              cell.barcode total.reads  cluster     gRNA1     gRNA2 sample.id</span></span>
<span><span class="co">#&gt;                               &lt;character&gt;   &lt;numeric&gt; &lt;factor&gt; &lt;numeric&gt; &lt;numeric&gt;  &lt;factor&gt;</span></span>
<span><span class="co">#&gt; AACAACCTACAATGTGCT-1 AACAACCTACAATGTGCT-1       27595     RPE1        18         0     gRNA1</span></span>
<span><span class="co">#&gt; AACAACTGGCGACCATCA-1 AACAACTGGCGACCATCA-1       33325     HCEC         0        66     gRNA2</span></span>
<span><span class="co">#&gt; AACAATGCAAGGATGCGT-1 AACAATGCAAGGATGCGT-1       36978     HCEC         0         0     NA   </span></span>
<span><span class="co">#&gt; AACAATGCATCGACGTTG-1 AACAATGCATCGACGTTG-1       25756     HCEC         0        34     gRNA2</span></span>
<span><span class="co">#&gt; AACAATGCATGAGAATCC-1 AACAATGCATGAGAATCC-1       26980     HCEC         0         0     NA   </span></span>
<span><span class="co">#&gt; ...                                   ...         ...      ...       ...       ...       ...</span></span>
<span><span class="co">#&gt; TTGGACTTCTATCATGCT-1 TTGGACTTCTATCATGCT-1       37431     HCEC         0         0     NA   </span></span>
<span><span class="co">#&gt; TTGGAGAACACGCAAGAT-1 TTGGAGAACACGCAAGAT-1       17456     RPE1         2         0     gRNA1</span></span>
<span><span class="co">#&gt; TTGGTAACTACCACTAGG-1 TTGGTAACTACCACTAGG-1       69296     RPE1       150         0     gRNA1</span></span>
<span><span class="co">#&gt; TTGGTAACTCAATCTCCG-1 TTGGTAACTCAATCTCCG-1       17866     RPE1         5         0     gRNA1</span></span>
<span><span class="co">#&gt; TTGTCAACCAGACTTCGT-1 TTGTCAACCAGACTTCGT-1       31021     HPNE         0         0     NA</span></span></code></pre></div>
<p>You can visualize the results a few different ways, including as the
annotations on a UMAP or heatmap. Here we show that each of the gRNAs
were transduced in a specific cell line from the experiment.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/reducedDimPlot.html">reducedDimPlot</a></span><span class="op">(</span><span class="va">tapexp</span>, dim.reduction <span class="op">=</span> <span class="st">"umap"</span>, group.label <span class="op">=</span> <span class="st">"sample.id"</span><span class="op">)</span></span></code></pre></div>
<p><img src="figure/barcode%20plots-1.png" width="100%" height="60%"></p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/assayHeatmap.html">assayHeatmap</a></span><span class="op">(</span><span class="va">tapexp</span>, alt.exp <span class="op">=</span> <span class="st">"smoothedCopyNumberByChr"</span>, assay <span class="op">=</span> <span class="st">"gmmCopyNumber"</span>,</span>
<span>    split.row.by <span class="op">=</span> <span class="st">"cluster"</span>, annotate.row.by <span class="op">=</span> <span class="st">"sample.id"</span>, color.preset <span class="op">=</span> <span class="st">"copy.number"</span><span class="op">)</span></span></code></pre></div>
<p><img src="figure/barcode%20plots-2.png" width="100%" height="60%"></p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">sessioninfo</span><span class="fu">::</span><span class="fu"><a href="https://sessioninfo.r-lib.org/reference/session_info.html" class="external-link">session_info</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="co">#&gt; ─ Session info ───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────</span></span>
<span><span class="co">#&gt;  setting  value</span></span>
<span><span class="co">#&gt;  version  R version 4.3.1 (2023-06-16)</span></span>
<span><span class="co">#&gt;  os       macOS Ventura 13.5</span></span>
<span><span class="co">#&gt;  system   x86_64, darwin20</span></span>
<span><span class="co">#&gt;  ui       RStudio</span></span>
<span><span class="co">#&gt;  language (EN)</span></span>
<span><span class="co">#&gt;  collate  en_US.UTF-8</span></span>
<span><span class="co">#&gt;  ctype    en_US.UTF-8</span></span>
<span><span class="co">#&gt;  tz       America/New_York</span></span>
<span><span class="co">#&gt;  date     2023-09-16</span></span>
<span><span class="co">#&gt;  rstudio  2023.06.0+421 Mountain Hydrangea (desktop)</span></span>
<span><span class="co">#&gt;  pandoc   3.1.1 @ /Applications/RStudio.app/Contents/Resources/app/quarto/bin/tools/ (via rmarkdown)</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ─ Packages ───────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────</span></span>
<span><span class="co">#&gt;  !  package              * version    date (UTC) lib source</span></span>
<span><span class="co">#&gt;     abind                  1.4-5      2016-07-21 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     askpass                1.2.0      2023-09-03 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     Biobase              * 2.60.0     2023-05-11 [1] Bioconductor</span></span>
<span><span class="co">#&gt;     BiocGenerics         * 0.46.0     2023-05-11 [1] Bioconductor</span></span>
<span><span class="co">#&gt;     BiocParallel           1.34.2     2023-05-22 [1] Bioconductor</span></span>
<span><span class="co">#&gt;     Biostrings             2.68.1     2023-05-16 [1] Bioconductor</span></span>
<span><span class="co">#&gt;     bitops                 1.0-7      2021-04-24 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     brio                   1.1.3      2021-11-30 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     bslib                  0.5.0      2023-06-09 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     cachem                 1.0.8      2023-05-01 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     Cairo                  1.6-0      2022-07-05 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     callr                  3.7.3      2022-11-02 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     circlize               0.4.15     2022-05-10 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     cli                    3.6.1      2023-03-23 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     clue                   0.3-64     2023-01-31 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     cluster                2.1.4      2022-08-22 [1] CRAN (R 4.3.1)</span></span>
<span><span class="co">#&gt;     codetools              0.2-19     2023-02-01 [1] CRAN (R 4.3.1)</span></span>
<span><span class="co">#&gt;     colorspace             2.1-0      2023-01-23 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     commonmark             1.9.0      2023-03-17 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     ComplexHeatmap         2.16.0     2023-05-11 [1] Bioconductor</span></span>
<span><span class="co">#&gt;     cowplot                1.1.1      2020-12-30 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     crayon                 1.5.2      2022-09-29 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     credentials            1.3.2      2021-11-29 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     curl                   5.0.2      2023-08-14 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     dbscan                 1.1-11     2022-10-27 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     DelayedArray           0.26.7     2023-07-28 [1] Bioconductor</span></span>
<span><span class="co">#&gt;     desc                   1.4.2      2022-09-08 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     devtools             * 2.4.5      2022-10-11 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     digest                 0.6.33     2023-07-07 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     distributional         0.3.2      2023-03-22 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     doParallel             1.0.17     2022-02-07 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     downlit                0.4.3      2023-06-29 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     dplyr                * 1.1.3      2023-09-03 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     ellipsis               0.3.2      2021-04-29 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     evaluate               0.21       2023-05-05 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     fansi                  1.0.4      2023-01-22 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     farver                 2.1.1      2022-07-06 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     fastmap                1.1.1      2023-02-24 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     fitdistrplus         * 1.1-11     2023-04-25 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     forcats              * 1.0.0      2023-01-29 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     foreach                1.5.2      2022-02-02 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     formatR                1.14       2023-01-17 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     fs                     1.6.3      2023-07-20 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     generics               0.1.3      2022-07-05 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     GenomeInfoDb         * 1.36.2     2023-08-25 [1] Bioconductor</span></span>
<span><span class="co">#&gt;     GenomeInfoDbData       1.2.10     2023-07-09 [1] Bioconductor</span></span>
<span><span class="co">#&gt;     GenomicRanges        * 1.52.0     2023-05-11 [1] Bioconductor</span></span>
<span><span class="co">#&gt;     gert                   1.9.2      2022-12-05 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     GetoptLong             1.0.5      2020-12-15 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     ggdist               * 3.3.0      2023-05-13 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     ggplot2              * 3.4.3      2023-08-14 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     ggrepel                0.9.3      2023-02-03 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     ggridges               0.5.4      2022-09-26 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     gh                     1.4.0      2023-02-22 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     gitcreds               0.1.2      2022-09-08 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     GlobalOptions          0.1.2      2020-06-10 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     glue                   1.6.2      2022-02-24 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     gtable                 0.3.4      2023-08-21 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     gtools                 3.9.4      2022-11-27 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     here                   1.0.1      2020-12-13 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     highr                  0.10       2022-12-22 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     htmltools              0.5.5      2023-03-23 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     htmlwidgets            1.6.2      2023-03-17 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     httpuv                 1.6.11     2023-05-11 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     httr2                  0.2.3      2023-05-08 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     IRanges              * 2.34.1     2023-06-22 [1] Bioconductor</span></span>
<span><span class="co">#&gt;     iterators              1.0.14     2022-02-05 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     janitor                2.2.0      2023-02-02 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     jquerylib              0.1.4      2021-04-26 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     jsonlite               1.8.7      2023-06-29 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;  VP karyotapR            * 1.0.1.9000 2023-09-07 [?] CRAN (R 4.3.1) (on disk 1.0.1)</span></span>
<span><span class="co">#&gt;     knitr                  1.43       2023-05-25 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     labeling               0.4.3      2023-08-29 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     later                  1.3.1      2023-05-02 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     lattice                0.21-8     2023-04-05 [1] CRAN (R 4.3.1)</span></span>
<span><span class="co">#&gt;     lifecycle              1.0.3      2022-10-07 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     lubridate              1.9.2      2023-02-10 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     magick                 2.7.4      2023-03-09 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     magrittr               2.0.3      2022-03-30 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     MASS                 * 7.3-60     2023-05-04 [1] CRAN (R 4.3.1)</span></span>
<span><span class="co">#&gt;     Matrix                 1.6-0      2023-07-08 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     MatrixGenerics       * 1.12.3     2023-07-31 [1] Bioconductor</span></span>
<span><span class="co">#&gt;     matrixStats          * 1.0.0      2023-06-02 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     memoise                2.0.1      2021-11-26 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     mime                   0.12       2021-09-28 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     miniUI                 0.1.1.1    2018-05-18 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     munsell                0.5.0      2018-06-12 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     openssl                2.1.0      2023-07-15 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     pillar                 1.9.0      2023-03-22 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     pkgbuild               1.4.2      2023-06-26 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     pkgconfig              2.0.3      2019-09-22 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     pkgdown                2.0.7      2022-12-14 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     pkgload                1.3.2.1    2023-07-08 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     png                    0.1-8      2022-11-29 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     prettyunits            1.1.1      2020-01-24 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     processx               3.8.2      2023-06-30 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     profvis                0.3.8      2023-05-02 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     promises               1.2.0.1    2021-02-11 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     ps                     1.7.5      2023-04-18 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     purrr                * 1.0.2      2023-08-10 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     R6                     2.5.1      2021-08-19 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     rappdirs               0.3.3      2021-01-31 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     rcmdcheck              1.4.0      2021-09-27 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     RColorBrewer           1.1-3      2022-04-03 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     Rcpp                   1.0.11     2023-07-06 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     RCurl                  1.98-1.12  2023-03-27 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     remotes                2.4.2.1    2023-07-18 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     reticulate             1.31       2023-08-10 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     rhdf5                  2.44.0     2023-05-11 [1] Bioconductor</span></span>
<span><span class="co">#&gt;     rhdf5filters           1.12.1     2023-05-11 [1] Bioconductor</span></span>
<span><span class="co">#&gt;     Rhdf5lib               1.22.0     2023-05-11 [1] Bioconductor</span></span>
<span><span class="co">#&gt;     rjson                  0.2.21     2022-01-09 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     rlang                  1.1.1      2023-04-28 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     rmarkdown              2.23       2023-07-01 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     roxygen2               7.2.3      2022-12-08 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     rprojroot              2.0.3      2022-04-02 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     Rsamtools              2.16.0     2023-05-11 [1] Bioconductor</span></span>
<span><span class="co">#&gt;     RSpectra               0.16-1     2022-04-24 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     rstudioapi             0.15.0     2023-07-07 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     S4Arrays               1.0.6      2023-08-30 [1] Bioconductor</span></span>
<span><span class="co">#&gt;     S4Vectors            * 0.38.1     2023-05-11 [1] Bioconductor</span></span>
<span><span class="co">#&gt;     sass                   0.4.7      2023-07-15 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     scales                 1.2.1      2022-08-20 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     sessioninfo            1.2.2      2021-12-06 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     shape                  1.4.6      2021-05-19 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     shiny                  1.7.4.1    2023-07-06 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     SingleCellExperiment * 1.22.0     2023-05-11 [1] Bioconductor</span></span>
<span><span class="co">#&gt;     snakecase              0.11.0     2019-05-25 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     stringdist             0.9.10     2022-11-07 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     stringi                1.7.12     2023-01-11 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     stringr                1.5.0      2022-12-02 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     SummarizedExperiment * 1.30.2     2023-06-06 [1] Bioconductor</span></span>
<span><span class="co">#&gt;     survival             * 3.5-5      2023-03-12 [1] CRAN (R 4.3.1)</span></span>
<span><span class="co">#&gt;     sys                    3.4.2      2023-05-23 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     testthat               3.1.10     2023-07-06 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     tibble                 3.2.1      2023-03-20 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     tidyr                  1.3.0      2023-01-24 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     tidyselect             1.2.0      2022-10-10 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     timechange             0.2.0      2023-01-11 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     umap                   0.2.10.0   2023-02-01 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     urlchecker             1.0.1      2021-11-30 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     usethis              * 2.2.2      2023-07-06 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     utf8                   1.2.3      2023-01-31 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     vctrs                  0.6.3      2023-06-14 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     viridisLite            0.4.2      2023-05-02 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     whisker                0.4.1      2022-12-05 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     withr                  2.5.0      2022-03-03 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     xfun                   0.39       2023-04-20 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     xml2                   1.3.5      2023-07-06 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     xopen                  1.0.0      2018-09-17 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     xtable                 1.8-4      2019-04-21 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     XVector                0.40.0     2023-05-11 [1] Bioconductor</span></span>
<span><span class="co">#&gt;     yaml                   2.3.7      2023-01-23 [1] CRAN (R 4.3.0)</span></span>
<span><span class="co">#&gt;     zlibbioc               1.46.0     2023-05-11 [1] Bioconductor</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  [1] /Library/Frameworks/R.framework/Versions/4.3-x86_64/Resources/library</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;  V ── Loaded and on-disk version mismatch.</span></span>
<span><span class="co">#&gt;  P ── Loaded and on-disk path mismatch.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────</span></span></code></pre></div>
  </main>
</div>



    <footer><div class="pkgdown-footer-left">
  <p>Developed by Joseph Mays.</p>
</div>

<div class="pkgdown-footer-right">
  <p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.1.3.</p>
</div>

    </footer>
</div>





  </body>
</html>
